src/models/model.cpp

```diff
@@ -249,32 +249,159 @@ int32_t Tokenizer::TokenToTokenId(const char* token) const {
+      for (auto& option : provider_options.options) {
+        opts.emplace(option.first, option.second);
+      }
+      session_options.AppendExecutionProvider("WebGPU", opts);
+    } else
+      throw std::runtime_error("Unknown provider type: " + provider_options.name);
+  }
+  return p_device;
+}
+
+// Trivial ONNX model that just returns a single float constant. Used below to create an OrtSession that
+// lets us get a device Ort::Allocator for each device type. This is necessary because the Ort::Allocator
+// needs to persist and is valid for the lifetime of this OrtSession.
+static const uint8_t g_trivial_model[] = {
+    0x08, 0x0a, 0x12, 0x01, 0x61, 0x3a, 0x53, 0x0a, 0x38, 0x12, 0x06, 0x76, 0x61, 0x6c, 0x75, 0x65,
+    0x73, 0x22, 0x08, 0x43, 0x6f, 0x6e, 0x73, 0x74, 0x61, 0x6e, 0x74, 0x2a, 0x24, 0x0a, 0x05, 0x76,
+    0x61, 0x6c, 0x75, 0x65, 0x2a, 0x18, 0x08, 0x01, 0x10, 0x01, 0x42, 0x0c, 0x63, 0x6f, 0x6e, 0x73,
+    0x74, 0x5f, 0x74, 0x65, 0x6e, 0x73, 0x6f, 0x72, 0x4a, 0x04, 0x00, 0x00, 0x00, 0x00, 0xa0, 0x01,
+    0x04, 0x12, 0x01, 0x62, 0x62, 0x14, 0x0a, 0x06, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x73, 0x12, 0x0a,
+    0x0a, 0x08, 0x08, 0x01, 0x12, 0x04, 0x0a, 0x02, 0x08, 0x01, 0x42, 0x04, 0x0a, 0x00, 0x10, 0x15};
+
 // Since Python/Others can and will hold onto a generator object past the model object's lifetime we need to ensure
 // the allocator used is not destroyed until last. This keeps the allocator around until exit, after all other memory
 // has been destroyed. Without this, we will crash in the Onnxruntime BFCArena code when deleting tensors due to the
 // arena already being destroyed.
-void EnsureDeviceOrtInit(OrtSession& session, DeviceType type) {
+void EnsureDeviceOrtInit(DeviceType type) {
   // CPU Allocator is a special case, it's not in the owned 'allocator_device_' table below so we handle it separately
   if (type == DeviceType::CPU)
     return;
 
-  auto& device = GetOrtGlobals()->allocator_device_[static_cast<int>(type)];
-  if (device)
+  auto& allocator = GetOrtGlobals()->device_allocators_[static_cast<int>(type)];
+  if (allocator.allocator_)
     return;
 
+  auto session_options = OrtSessionOptions::Create();
+
+  const char* device_name{};
+  switch (type) {
+    case DeviceType::CUDA:
+      device_name = "cuda";
+      break;
+    case DeviceType::DML:
+      device_name = "dml";
+      break;
+    case DeviceType::WEBGPU:
+      device_name = "webgpu";
+      break;
+    case DeviceType::QNN:
+      device_name = "qnn";
+      break;
+    default:
+      throw std::runtime_error("Unknown device type");
+  }
+
+  std::vector<Config::ProviderOptions> provider_options_list;
+  provider_options_list.emplace_back(Config::ProviderOptions{device_name, {}});
+
+  SetProviderSessionOptions(*session_options, provider_options_list, true, false);
+  allocator.session_ = OrtSession::Create(GetOrtEnv(), g_trivial_model, sizeof(g_trivial_model), session_options.get());
```

Comment:
A comment explaining the setup to the new reader would be great so they don't have to infer 'why' we're doing all this. 

We create a session with the required EP. We put that session into the globals so it remains valid for the lifetime of GenAI. Due to that the allocator we get from that session is guaranteed to remain valid and can be used for any other models we load.